File: backend/app/schemas/reports.py
"""Pydantic response schemas for the Phase 10 Reports module.

Monetary fields are exposed as ``float`` to match the convention established in
[backend/app/schemas/finance.py](backend/app/schemas/finance.py) (Phase 8) — the
frontend TypeScript layer is already typed against numbers, and JSON-encoded
Decimals would otherwise come through as strings.
"""

from datetime import date
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Dashboard
# ---------------------------------------------------------------------------


class SparklinePoint(BaseModel):
    date: date
    value: float


class DashboardTopProduct(BaseModel):
    product_id: UUID
    product_name: str
    units_sold: int


class DashboardActivityItem(BaseModel):
    kind: str  # "sale" | "production" | "purchase_order" | "wastage"
    summary: str
    occurred_at: date
    actor: Optional[str] = None


class DashboardResponse(BaseModel):
    """Adapts to the requesting user's role.

    Fields the user is not authorised to see are returned as null rather
    than being omitted (keeps the TypeScript surface stable across roles).
    """

    role: str

    # KPIs — financial (Owner, Finance Manager, Store Manager)
    revenue_today: Optional[float] = None
    revenue_week: Optional[float] = None
    revenue_month: Optional[float] = None
    gross_profit_week: Optional[float] = None
    units_sold_today: Optional[int] = None

    # KPIs — operational (Owner, Production Manager)
    production_batches_today: Optional[int] = None
    active_stock_alerts: Optional[int] = None
    expiring_ingredients: Optional[int] = None
    pending_purchase_orders: Optional[int] = None

    # Trend data — last 7 days
    revenue_sparkline: List[SparklinePoint] = []
    batches_sparkline: List[SparklinePoint] = []

    # Spotlight
    top_product_today: Optional[DashboardTopProduct] = None

    # Recent activity (latest 5 across sales/POs/batches/wastage)
    recent_activity: List[DashboardActivityItem] = []

    # Context echoed back so the UI can label everything correctly
    scoped_store_id: Optional[UUID] = None
    scoped_store_name: Optional[str] = None


# ---------------------------------------------------------------------------
# Sales trends + Top sellers
# ---------------------------------------------------------------------------


class SalesTrendPoint(BaseModel):
    date: date
    units_sold: int
    revenue: float
    transaction_count: int


class SalesTrendsResponse(BaseModel):
    granularity: str  # "day" | "week"
    points: List[SalesTrendPoint] = []
    total_units: int
    total_revenue: float


class TopSellerItem(BaseModel):
    product_id: UUID
    product_name: str
    sku: Optional[str] = None
    units_sold: int
    revenue: float
    avg_unit_price: float


class TopSellersResponse(BaseModel):
    order_by: str  # "units" | "revenue"
    items: List[TopSellerItem] = []


# ---------------------------------------------------------------------------
# Wastage trends
# ---------------------------------------------------------------------------


class WastageTrendBucket(BaseModel):
    """Generic bucket — `key` is a string regardless of group_by so the UI
    can render all three modes uniformly (date / reason / source)."""

    key: str
    label: str  # human-friendly version of `key`
    total_qty: int
    total_cost: float
    record_count: int


class WastageTrendsResponse(BaseModel):
    group_by: str  # "date" | "reason" | "source"
    buckets: List[WastageTrendBucket] = []
    total_qty: int
    total_cost: float


# ---------------------------------------------------------------------------
# Ingredient consumption
# ---------------------------------------------------------------------------


class IngredientConsumptionItem(BaseModel):
    ingredient_id: UUID
    ingredient_name: str
    unit: str
    total_qty_consumed: float
    total_cost: float
    batch_count: int


class IngredientConsumptionResponse(BaseModel):
    items: List[IngredientConsumptionItem] = []
    total_cost: float


# ---------------------------------------------------------------------------
# Production efficiency
# ---------------------------------------------------------------------------


class ProductionByRecipeItem(BaseModel):
    recipe_id: UUID
    recipe_name: str
    planned_batches: int
    completed_batches: int
    cancelled_batches: int
    total_planned_qty: int
    total_actual_qty: int
    avg_yield_variance_pct: float


class ProductionEfficiencyResponse(BaseModel):
    planned_count: int
    in_progress_count: int
    completed_count: int
    cancelled_count: int
    total_batches: int
    completion_rate: float
    avg_yield_variance_pct: float
    by_recipe: List[ProductionByRecipeItem] = []

File: frontend/src/api/reports.ts
import apiClient from './client';

// ---------------------------------------------------------------------------
// Dashboard
// ---------------------------------------------------------------------------

export interface SparklinePoint {
  date: string;
  value: number;
}

export interface DashboardTopProduct {
  product_id: string;
  product_name: string;
  units_sold: number;
}

export interface DashboardActivityItem {
  kind: 'sale' | 'production' | 'purchase_order' | 'wastage';
  summary: string;
  occurred_at: string;
  actor?: string | null;
}

export interface DashboardResponse {
  role: string;
  revenue_today: number | null;
  revenue_week: number | null;
  revenue_month: number | null;
  gross_profit_week: number | null;
  units_sold_today: number | null;
  production_batches_today: number | null;
  active_stock_alerts: number | null;
  expiring_ingredients: number | null;
  pending_purchase_orders: number | null;
  revenue_sparkline: SparklinePoint[];
  batches_sparkline: SparklinePoint[];
  top_product_today: DashboardTopProduct | null;
  recent_activity: DashboardActivityItem[];
  scoped_store_id: string | null;
  scoped_store_name: string | null;
}

// ---------------------------------------------------------------------------
// Sales trends + top sellers
// ---------------------------------------------------------------------------

export interface SalesTrendPoint {
  date: string;
  units_sold: number;
  revenue: number;
  transaction_count: number;
}

export interface SalesTrendsResponse {
  granularity: 'day' | 'week';
  points: SalesTrendPoint[];
  total_units: number;
  total_revenue: number;
}

export interface TopSellerItem {
  product_id: string;
  product_name: string;
  sku?: string | null;
  units_sold: number;
  revenue: number;
  avg_unit_price: number;
}

export interface TopSellersResponse {
  order_by: 'units' | 'revenue';
  items: TopSellerItem[];
}

// ---------------------------------------------------------------------------
// Wastage trends
// ---------------------------------------------------------------------------

export interface WastageTrendBucket {
  key: string;
  label: string;
  total_qty: number;
  total_cost: number;
  record_count: number;
}

export interface WastageTrendsResponse {
  group_by: 'date' | 'reason' | 'source';
  buckets: WastageTrendBucket[];
  total_qty: number;
  total_cost: number;
}

// ---------------------------------------------------------------------------
// Ingredient consumption
// ---------------------------------------------------------------------------

export interface IngredientConsumptionItem {
  ingredient_id: string;
  ingredient_name: string;
  unit: string;
  total_qty_consumed: number;
  total_cost: number;
  batch_count: number;
}

export interface IngredientConsumptionResponse {
  items: IngredientConsumptionItem[];
  total_cost: number;
}

// ---------------------------------------------------------------------------
// Production efficiency
// ---------------------------------------------------------------------------

export interface ProductionByRecipeItem {
  recipe_id: string;
  recipe_name: string;
  planned_batches: number;
  completed_batches: number;
  cancelled_batches: number;
  total_planned_qty: number;
  total_actual_qty: number;
  avg_yield_variance_pct: number;
}

export interface ProductionEfficiencyResponse {
  planned_count: number;
  in_progress_count: number;
  completed_count: number;
  cancelled_count: number;
  total_batches: number;
  completion_rate: number;
  avg_yield_variance_pct: number;
  by_recipe: ProductionByRecipeItem[];
}

// ---------------------------------------------------------------------------
// Filter types
// ---------------------------------------------------------------------------

export interface DateRangeFilters {
  date_from?: string;
  date_to?: string;
}

export interface SalesTrendFilters extends DateRangeFilters {
  store_id?: string;
  product_id?: string;
  granularity?: 'day' | 'week';
}

export interface TopSellersFilters extends DateRangeFilters {
  store_id?: string;
  limit?: number;
  order_by?: 'units' | 'revenue';
}

export interface WastageTrendFilters extends DateRangeFilters {
  store_id?: string;
  group_by?: 'date' | 'reason' | 'source';
}

export interface IngredientConsumptionFilters extends DateRangeFilters {
  ingredient_id?: string;
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

export const reportsApi = {
  getDashboard: async (): Promise<DashboardResponse> => {
    const response = await apiClient.get('/reports/dashboard');
    return response.data;
  },

  getSalesTrends: async (
    params?: SalesTrendFilters
  ): Promise<SalesTrendsResponse> => {
    const response = await apiClient.get('/reports/sales-trends', { params });
    return response.data;
  },

  getTopSellers: async (
    params?: TopSellersFilters
  ): Promise<TopSellersResponse> => {
    const response = await apiClient.get('/reports/top-sellers', { params });
    return response.data;
  },

  getWastageTrends: async (
    params?: WastageTrendFilters
  ): Promise<WastageTrendsResponse> => {
    const response = await apiClient.get('/reports/wastage-trends', { params });
    return response.data;
  },

  getIngredientConsumption: async (
    params?: IngredientConsumptionFilters
  ): Promise<IngredientConsumptionResponse> => {
    const response = await apiClient.get('/reports/ingredient-consumption', {
      params,
    });
    return response.data;
  },

  getProductionEfficiency: async (
    params?: DateRangeFilters
  ): Promise<ProductionEfficiencyResponse> => {
    const response = await apiClient.get('/reports/production-efficiency', {
      params,
    });
    return response.data;
  },
};
